### PR TITLE
The tree construction algorithm

### PR DESCRIPTION
**Pull request by bobg**
This description of the algorithm for building a hashsplit tree is based on the existing implementation at https://github.com/bobg/hashsplit.

### CODE REVIEW

**Comment by bobg** [comment]
Apologies to @cole-miller. I am happy to defer to your work in https://github.com/hashsplit/hashsplit-spec/pull/12 if you intend to finish it, but it's been a few weeks.

**Comment by cole-miller** [comment]
No, you're quite right! Thanks for taking this on.

File: spec.md [hunk 1]
```diff
@@ -105,7 +105,73 @@ We define $\operatorname{SPLIT}_C(X)$ recursively, as follows:
...
+The trees representing two slightly different versions of the same input
+will differ only in the subtrees in the vicinity of the differences.
+
+## Definitions
+
+The “level” $L(X)$ of a sequence $X$ is $Q - T$,
+where $Q$ is the largest integer such that
+
+- $Q <= 32$ and
+- $H(P(X)) \mod 2^Q = 0$
```
Comment (zenhack) [inline spec.md]: This isn't quite right; `H` takes the sequence of bytes to compute the hash over, so it should be just the last `W` bytes of `P(X)`.

Also, you should probably talk explicitly about the case where the split is triggered by hitting S_max or the end of input (and thus `Q < T`)

File: spec.md [hunk 2]
```diff
@@ -105,7 +105,73 @@ We define $\operatorname{SPLIT}_C(X)$ recursively, as follows:
...
+is determined by the content of the input.
+The trees representing two slightly different versions of the same input
+will differ only in the subtrees in the vicinity of the differences.
+
+## Definitions
+
+The “level” $L(X)$ of a sequence $X$ is $Q - T$,
+where $Q$ is the largest integer such that
+
+- $Q <= 32$ and
```
Comment (zenhack) [inline spec.md]: Should use `\le` here, so it renders correctly.

File: spec.md [hunk 3]
```diff
@@ -105,7 +105,73 @@ We define $\operatorname{SPLIT}_C(X)$ recursively, as follows:
...
+- $H(P(X)) \mod 2^Q = 0$
+
+(i.e., the level is the number of trailing zeroes in the rolling checksum in excess of the threshold needed to produce the prefix chunk $P(X)$).
+
+A “node” in a hashsplit tree
+is a pair $\langle D, C \rangle$
```
Comment (zenhack) [inline spec.md]: I think we should use a different notation for tuples like this than for sequences; maybe parens?

File: spec.md [hunk 4]
```diff
@@ -105,7 +105,73 @@ We define $\operatorname{SPLIT}_C(X)$ recursively, as follows:
...
+    d. Terminate with $N_d$ as the root node.
+3. Otherwise, $N_0 \leftarrow \langle 0, \operatorname{Children}(N_0) \mathbin{|} \langle P(X) \rangle \rangle$ (i.e., add $P(X)$ to the list of children in $N_0$).
+4. For each integer $i$ in $[0 .. L(X))$, “close” the node $N_i$ (see below).
+5. $X \leftarrow R(X)$.
+6. Go to step 2.
+
+To “close” a node $N_i$:
+
+1. If no $N_{i+1}$ exists yet, let $N_{i+1}$ be $\langle i+1, \langle\rangle \rangle$ (i.e., a node at depth ${i + 1}$ with no children).
+2. $N_{i+1} \leftarrow \langle i+1, \operatorname{Children}(N_{i+1}) \mathbin{|} \langle N_i \rangle \rangle$ (i.e., add $N_i$ as a child to $N_{i+1}$).
```
Comment (zenhack) [inline spec.md]: You want `\mathbin{\|}` (notice the extra `\`); this is rendering as a single pipe currently.

**Review by zenhack — COMMENTED** [review]
A few minor things in-line.

Additionally, I feel like there ought to be a way to formulate this that mirrors SPLIT; there is more symmetry than is immediately apparent given how different the presentations are currently, and with this description there's a lot of state to keep in one's head -- I think we can probably do better for a description of "what" is being computed; perhaps if we feel it is useful the "how" of the imperative description could stick around in an "implementation" section, like we have with the rrs1 hash function.

**Comment by bobg** [comment]
> I feel like there ought to be a way to formulate this that mirrors SPLIT

I have an idea of how to do a more algebraic (as opposed to procedural) description of TREE_C(X), but not a terribly clear one yet. How would you feel about merging the current change and tackling this in a followup PR?

**Comment by zenhack** [comment]
Works for me; looking forward to the followup.

**Comment by bobg** [comment]
Groovy, thanks. Preview of work in progress here: https://github.com/hashsplit/hashsplit-spec/compare/master...bobg:algebraic-tree

**Pull request merged**